delivery_carrier_label_batch/wizard/generate_labels.py

```diff
@@ -0,0 +1,282 @@
+# -*- coding: utf-8 -*-
+# Copyright 2013-2016 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl)
+import Queue
+import logging
+import openerp
+import threading
+from contextlib import contextmanager
+from itertools import groupby
+from openerp import _, api, exceptions, fields, models
+
+from ..pdf_utils import assemble_pdf
+from ..zpl_utils import assemble_zpl2
+
+_logger = logging.getLogger(__name__)
+
+
+class DeliveryCarrierLabelGenerate(models.TransientModel):
+
+    _name = 'delivery.carrier.label.generate'
+
+    @api.multi
+    def _get_batch_ids(self):
+        res = False
+        if (self.env.context.get('active_model') == 'stock.batch.picking' and
+                self.env.context.get('active_ids')):
+            res = self.env.context['active_ids']
+        return res
+
+    batch_ids = fields.Many2many(
+        'stock.batch.picking',
+        string='Picking Batch',
+        default=_get_batch_ids)
+    generate_new_labels = fields.Boolean(
+        'Generate new labels',
+        default=False,
+        help="If this option is used, new labels will be "
+             "generated for the packs even if they already have one.\n"
+             "The default is to use the existing label.")
+
+    @api.model
+    def _get_packs(self, batch):
+        operations = batch.pack_operation_ids
+        operations = sorted(
+            operations,
+            key=lambda r: r.result_package_id.name or r.package_id.name
+        )
+        for pack, grp_operations in groupby(
+                operations,
+                key=lambda r: r.result_package_id or r.package_id):
+            pack_label = self._find_pack_label(pack)
+            yield pack, list(grp_operations), pack_label
+
+    @api.model
+    def _find_picking_label(self, picking, f_type=None):
+        label_obj = self.env['shipping.label']
+        domain_file_type = ('file_type', '=', f_type)
+        if not f_type:
+            domain_file_type = ('file_type', 'in', self.get_file_types())
```

Comment:
I thought file_type is mandatory, anyhow if f_type=None doesn't this meant that we should search for all files? do  we need to call get_file_types here?

fyi in v12 file_type was changed from selection to char field, maybe worth to mention in code when will work on migration and writing tests

Reply:
There are existing calls to this method IIRC (outside of this module) that are not aware of the file type, so we need to backward-compatible. :+1: to avoid a call to `get_file_types()`.